includes/query/SMW_QueryProcessor.php

```diff
@@ -52,26 +53,35 @@ public static function setRecursiveTextProcessor( RecursiveTextProcessor $recurs
 	 * param name (string) => param value (mixed)
 	 *
 	 * @since 1.6.2
-	 * The return value changed in SMW 1.8 from an array with result values
-	 * to an array with Param objects.
+	 * The return value changed in SMW 3.1 from Param[] to ProcessedParam[]
 	 *
 	 * @param array $params
-	 * @param array $printRequests
+	 * @param PrintRequest[] $printRequests
 	 * @param boolean $unknownInvalid
 	 *
-	 * @return Param[]
+	 * @return ProcessedParam[]
 	 */
 	public static function getProcessedParams( array $params, array $printRequests = [], $unknownInvalid = true, $context = null, $showMode = false ) {
-		$validator = self::getValidatorForParams( $params, $printRequests, $unknownInvalid, $context, $showMode );
-		$validator->processParameters();
-		$parameters =  $validator->getParameters();
-
-		// Negates some weird behaviour of ParamDefinition::setDefault used in
-		// an individual printer.
-		// Applying $smwgQMaxLimit, $smwgQMaxInlineLimit will happen at a later
-		// stage.
+		$validator = self::getProcessorForParams( $params, $printRequests, $unknownInvalid, $context, $showMode );
+		$processingResult = $validator->processParameters();
+
+		$parameters = [];
+
+		foreach ( $processingResult->getParameters() as $parameter ) {
+			$parameters[$parameter->getName()] = $parameter;
+		}
+
+		// Set limit parameter to its original value.
+		// This is to allow for special handling of invalid values rather than just setting the parameter to its default.
+		// $smwgQMaxLimit and $smwgQMaxInlineLimit will be applied at a later point.
 		if ( isset( $params['limit'] ) && isset( $parameters['limit'] ) ) {
-			$parameters['limit']->setValue( (int)$params['limit'] );
+			$parameters['limit'] = new ProcessedParam(
```

Comment:
I felt particular uneasy about this alteration knowing the limit setting has caused a regression in the past therefore I devised an integration test #4187 that would fail if the limit where causing some unexpected changes due to use of `ParamDefinition::setDefault`.

Reply:
As expected from the Hero Of The Tests ;)